### DISCUSSION TITLE
Global Katex \newcommand

### DISCUSSION THREAD

**Discussion by fdegrave** [marp-team/marp#71] [Q&A] [answered]
First of all, thank you for making my teacher's life easier! I use Marp a lot for math notations, and I often use things like:

```
$$\newcommand{\cons}[2]{\mathtt{cons}\ #1\  #2}
...
$$
```
It works perfectly, but I need to redefine all my commands in each KaTeX environment, they are not "kept" for the rest of the document. Is there a way to define my `\newcommand` only once in the .md document, and use them anywhere after?

**Comment by yhatt** [comment]
Currently each KaTeX renderings in Marp Markdown have individual contexts so `\newcommand` is working only in a defined environment. Fortunately there is a way to [persist defined commands](https://github.com/KaTeX/KaTeX/blob/master/docs/api.md#persistent-macros) in KaTeX API, so we might move for that if received many requests.

### Workaround for Marp CLI

> Macros can also be defined in the KaTeX [rendering options.](https://katex.org/docs/options.html)
>
> _-- https://katex.org/docs/supported.html#macros_

Our core engine can [customize KaTeX rendering option](https://github.com/marp-team/marp-core#math-boolean--katex--mathjax--object), and Marp CLI can modify it by using [the configuration file](https://github.com/marp-team/marp-cli#advanced). Try using this configuration if you are using Marp CLI:

### `.marprc.json`

```json
{
  "options": {
    "math": {
      "katexOption": {
        "macros": {
          "\\cons": "\\mathtt{cons}\\ #1\\  #2"
        }
      }
    }
  }
}
```

**Reply by fdegrave** [reply]
Thank you! However I use Marp in VSCode, is there a way to provide the same option through the settings.json?

**Reply by yhatt** [reply]
No. Marp for VS Code has no way to set advanced option because [allowing the advanced configuration like that means to open a security hole](https://github.com/marp-team/marp-vscode/pull/123#issuecomment-691507780). We are responsible for saving users from any malicious. Please use Marp CLI.

**Comment by yhatt** [comment, marked as answer]
### [UPDATED]

Hey @fdegrave, I've updated Marp's math handling in marp-team/marp-core#212, to persist KaTeX macros, that have defined as global, between math typesetting in a Markdown. It has shipped as [Marp Core v1.4.1](https://github.com/marp-team/marp-core/releases/tag/v1.4.1), and that has bundled into the latest update of Marp tools, CLI [v0.23.1](https://github.com/marp-team/marp-cli/releases/tag/v0.23.1) and VS Code extension [v0.17.1](https://github.com/marp-team/marp-vscode/releases/tag/v0.17.1)!

KaTeX has supported following commands to define new command as globally:

> `\gdef`, `\xdef`, `\global\def`, `\global\edef`, `\global\let`, and `\global\futurelet` will persist between math expressions.
>
> _-- https://katex.org/docs/supported.html#macros_

As you can see from this quote, please notice that `\newcommand` you're using cannot persist the command.

**Reply by fdegrave** [reply]
Great! Thank you so much! Well noted for the unsupported `\newcommand`, I'll manage ;-)
